packages/opentelemetry-types/src/trace/Sampler.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SpanContext } from './span_context';
+
+/** This interface represent a sampler. */
+export interface Sampler {
+  /**
+   * A string that uniquely describes the sampling behavior of this instance.
+   */
+  readonly description: string;
+
+  /**
+   * Checks whether span needs to be created and tracked.
+   *
+   * @param traceId the TraceId for the new Span. This will be identical to
+   *     that in the parentContext, unless this is a root span.
+   * @param spanId the SpanId for the new Span.
+   * @param [name] the name of the new Span.
+   * @param [parentContext] Parent span context. Typically taken from the wire.
+   *     Can be null.
+   * @returns whether span should be sampled or not.
+   */
+  shouldSample(
```

Comment:
Here, only the span context needs to be passed. Then it should be possible to retrieve all of these directly from the context.

Reply:
`parentContext` is taken from the wire, which means if the parent is sampled keep the sampling decision or else fallback on `traceId` logic to make sampling decision. 

I think `spanId` (_of new span_) and `name` (_either service_name or span name_) are provided, if someone(or vender) wants to build their customized sampling decision based on these arguments.

I have to add another optional argument `parentLinks`, which are a collection of links that will be associated with the Span. The idea is, if any parent link is sampled keep the sampling decision. (_will include in same PR_).

Required arguments: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/sampling-api.md#shouldsample

Reply:
This API is very limiting. For example, as a vendor, if I only care about one argument, I am still forced to pass all of them. If I want to make a decision based on an argument that is not provided, I need to make a change to the OTel spec or break the contract.

Passing only a single span context for the new span, and exposing these other arguments directly on the span context would result in a much cleaner API and allow vendors to make a decision based on any property on the span context. This approach is also close enough to the spec since it's actually a cleaner superset.

Reply:
What do you think about having `SamplerOptions` with all these optional properties?

Something like:
```
export interface SamplerOptions {
    spanContext: SpanContext
    traceId?: string, 
    spanId?: string,
    name?: string,
    links?: Link[]
}

export interface Sampler {
   readonly description: string;
   shouldSample(samplerOptions?: SamplerOptions): boolean;
}
```

Reply:
This would definitely be a lot better, and more inline with how optional parameters are usually passed in JavaScript.

However, I still think it's an inferior solution than having everything available on the span context, which would have the following advantages:

* Possible to simply pass the span context and not have to worry about what data to pass. Tell don't ask. In my experience the `SpanContext` (at least in OpenTracing) makes this pattern very easy (basically you always tell the span/context and never need to ask).
* This is all useful data that could be needed in other places. This means the same logic to find this data and pass it explicitly will be required in many places.
* Whenever it's possible to pass a span context, it should also be possible to pass a span, which makes this even simpler in cases where what you have is actually a span. This point could actually be solved with the new proposed API as well, although in a less seamless way.

For example:

```js
// good
sampler.shouldSample(span)
sampler.shouldSample(spanContext)

// bad, trace ID is already available on the context and thus redundant
sampler.shouldSample({ spanContext, traceId })
```

Reply:
>  I still think it's an inferior solution than having everything available on the span context.

I will update the PR based on conclusion here https://github.com/open-telemetry/opentelemetry-js/pull/45#discussion_r296005681

Reply:
@rochdev - SpanContext is defined in the https://github.com/w3c/trace-context, and it contains the minimum necessary information that needs to be passed between parent and child spans (on the wire or inside the process).

All the extra information that is required to do sampling or other decisions should not be in the SpanContext but passed to these components via options or optional arguments.

Reply:
It sounds to me then that `SpanContext` is used specifically for distributed context propagation then, which is completely different than the `SpanContext` from OpenTracing.

If that's the case, it means that the internal context as is defined right now in OpenTelemetry is not the correct class. Whatever will be held by the span as its "context" only be converted to a `SpanContext` at propagation time. Whatever is held in the span must be mutable and allow extensions.

Should `span.context()` be renamed to avoid confusion?

Reply:
To summarize the issue better and what is needed:

* Vendors need a way to attach arbitrary data on the span, which in OpenTracing is called `SpanContext`
* This needs to be on every span so that it's propagated within the process.
* It needs to be mutable as this arbitrary data can be added/modified at any time depending on vendor needs.
* It shouldn't be serialized right away given it's mutable nature. This means that the W3C `SpanContext` should only exist just before/after propagation.
* The sampler should only require this object, which would contain all the information in the current signature, plus any other vendor specific data.

It doesn't need to be called `SpanContext` so it could be different. Maybe it even already exists and I missed it, please let me know if that's the case. I'm not super familiar with the spec since I'm staying clear of the spec repo for now as I'm otherwise overloaded by all the issues this API currently has, especially from a JavaScript perspective, so I'm trying to address each class as they come in this repo.